Consul concept updates (#1705)
hashicorp/web-unified-docs: content/consul/v1.22.x/content/docs/concept/reliability.mdx

To build a resilient platform, you must minimize the number of remediation actions you need to take when a failure occurs. Consul's core operations rely on server agents, which use the [Raft protocol for consensus](/consul/docs/concept/consensus) to provide results from the service catalog.

_Fault tolerance_ refers to a system's ability to continue operating without interruption, even when one or more components fail. In Consul, the number of server agents determines the cluster's fault tolerance.

## Fault tolerance

There are several ways to improve Consul's fault tolerance. For enhanced reliability, we recommend using multiple methods together.

- Use a [minimum quorum size](#quorum-size) to avoid performance impacts.
- Spread servers across infrastructure [availability zones](#availability-zones).
- Use [redundancy zones](#redundancy-zones) to improve fault tolerance. <EnterpriseAlert inline /> 
- Use [Autopilot](#autopilot) to automatically prune failed servers and maintain quorum size. <EnterpriseAlert inline /> 
- Use [cluster peering](#cluster-peering) to provide service redundancy.

## Quorum size

> The following table shows quorum size and failure tolerance for various cluster sizes. The recommended deployment for production is either 3 or 5 servers. We recommend limiting the use of datacenters that use a single server agent to development scenarios.

Reviewer: ```suggestion
The following table shows quorum size and failure tolerance for various cluster sizes. We recommend either 3 or 5 servers for production deployments. You should restrict the use of single server agent datacenters to development scenarios.
```

We recommend deploying clusters with either 3 or 5 voting server agents. At these sizes, the cluster is fault tolerant without becoming too large. Spreading voting servers across availability zones can further improve fault tolerance, if desired.

## Availability zones

The cloud or on-premise infrastructure underlying your Consul datacenter can run across multiple availability zones. 

An availability zone is meant to share no points of failure with other zones by:

- Having power, cooling, and networking systems independent from other zones
- Being physically distant enough from other zones so that large-scale disruptions
  such as natural disasters (flooding, earthquakes) are very unlikely to affect multiple zones

Availability zones are available in the regions of most cloud providers and in some on-premise installations.

If possible, spread your Consul voting servers across 3 availability zones to protect your Consul datacenter from a failure in any single availability zone. For example, if you deploy 5 Consul servers across 3 availability zones, limit each zone to 2 servers.

> If one zone fails, you will lose 2 servers, but quorum will be maintained by the 3 remaining servers. 

Reviewer: ```suggestion
If one zone fails, you lose two servers, but quorum is maintained by the three remaining servers. 
```

When all servers are healthy, only one server in the redundancy zone will be an active voter. All other servers will be backup voters.
If a zone's voter is lost, Consul replaces it with a backup voter in the same zone, if any. Otherwise, it replaces the missing voter with a backup voter within another zone.

In most cases, Consul can replace lost voters with backup voters within 30 seconds.

Because this replacement process is not instantaneous, redundancy zones do not improve _immediate fault tolerance_, which is
the number of healthy voting servers that can fail simultaneously without causing an outage.

Instead, redundancy zones improve _optimistic fault tolerance_, or the number of healthy active and back-up voting servers that can fail over time without causing an outage.

For example, consider a Consul datacenter with 3 redundancy zones, and 2 servers per zone. This cluster has the following characteristics:

- There are 6 Consul servers in the cluster.
- There are 3 voting servers.
- The quorum size is 2.
- The immediate fault tolerance is 1.
- The optimistic fault tolerance is 4, meaning 4 servers can fail over time before the cluster loses quorum.

> As a result, a Consul cluster running 6 total servers could perform similar to a 3 server datacenter, but still provide fault tolerance similar to a 7 server datacenter.

Reviewer: ```suggestion
- There are three voting servers.
- The quorum size is two.
- The immediate fault tolerance is one.
- The optimistic fault tolerance is four, meaning four servers can fail over time before the cluster loses quorum.

As a result, a Consul cluster running six total servers could perform similar to a three server datacenter, but still provide fault tolerance similar to a seven server datacenter.
```
personally I find numbers easier to read, but our style guide specifies using words for small numbers

Author: I'm going to declare this an exception to that rule. There are so many numbers that they make it easier to interpret the list.
